### ISSUE TITLE
[dev-server-esbuild] Unable to resolve packages with correct extensions when imported in an .ts file using node's `exports`

### ISSUE THREAD

**Issue by motss** [modernweb-dev/web#1376]
When files being imported in an `.ts` file, the resolver automatically assumes all the imports must be ended with `.ts` when some could be `.js`. It seems like the dev server does not support node's `exports` quite well.

## Actual behavior

Packages that use `exports` in node resolves with `.ts` when it gets imported in a `.ts` file.

## Expected behavior

Packages that use `exports` in node must be resolved with the correct extensions when it gets imported in a `.ts` file

## Steps to reproduce

Clone this [repo](https://github.com/motss/wds-import-bug) and run the project to reproduce the stated problem.

**Comment by LarsDenBakker** [comment]
I looked into this, and it's an issue inside rollup node resolve. There is special handling for .ts files: https://github.com/rollup/plugins/blob/master/packages/node-resolve/src/index.js#L178. Together with the export map of this package, it creates this situation...

**Comment by motss** [comment]
Not sure if it is a rollup plugin issue but I tried it with `rollup`, `rollup-plugin-typescript2` and `@rollup/plugin-node-resolve`. Everything runs without issue. There is no resolution issue with `node`'s export map.

**Comment by LarsDenBakker** [comment]
Will need to do some more investigation on our side

**Comment by peschee** [comment]
@LarsDenBakker @motss I think I am running into the same or similar issue when using Lit 2.x directives. I have created a small test repo: https://github.com/peschee/wds-esbuild-lit-test

**Comment by peschee** [comment]
I think this issue blocks users from upgrading to Lit 2.x when using TypeScript and the recommended esbuild setup in wds.

**Comment by swissmanu** [comment]
Same problem here, in conjunction with `lit` 2.x.

Am I right with my assumption, that we would have to change the behavior here? https://github.com/modernweb-dev/web/blob/master/packages/dev-server-esbuild/src/EsbuildPlugin.ts#L79-L88

**Comment by abdonrd** [comment]
> @LarsDenBakker @motss I think I am running into the same or similar issue when using Lit 2.x directives. I have created a small test repo: https://github.com/peschee/wds-esbuild-lit-test

I have the exactly same problem!

**Comment by abdonrd** [comment]
@peschee have you any workaround?

**Comment by abdonrd** [comment]
> Will need to do some more investigation on our side

Have you had time to invest it? Thanks in advance!

**Comment by abdonrd** [comment]
> Will need to do some more investigation on our side

@LarsDenBakker Have you had time to investigate it? Thanks in advance!

**Comment by thescientist13** [comment]
Not sure if related, but I may be running into something similar?  I'm using test runner, but as I understand it, plugins and middleware get handled by dev server, so I think the context applies?

Was using web test server, writing some ES2015+ custom elements and specs with new lit, everything is good.  💯 
https://github.com/AnalogStudiosRI/www.analogstudios.net/tree/chore/design-system-poc

Added a new TypeScript based component following the lit.dev playground 
https://github.com/AnalogStudiosRI/www.analogstudios.net/tree/chore/design-system-poc-typescript

Now getting this error when running test runner
```sh
src/components/greeting/greeting.spec.js:

 🚧 Browser logs:
      TypeError: Failed to resolve module specifier "lit". Relative references must start with either "/", "./", or "../".

 ❌ Could not import your test module. Check the browser logs or open the browser in debug mode for more information. 

Chrome: |██████████████████████████████| 3/3 test files | 6 passed, 0 failed

Code coverage: 100 %

Finished running tests, watching for file changes... 
 ```
 
I noticed comparing the responses using the browser debugger that Dev Server (presumably) was correctly re-writing my bare imports in my other components, like _footer.js_
<img width="1662" alt="Screen Shot 2021-07-04 at 11 30 10 AM" src="https://user-images.githubusercontent.com/895923/124391264-c2cbd500-dcbd-11eb-851e-471726aef0ff.png">

But not for _greeting.ts_.  Maybe that's because `import` is not at the top?
<img width="1671" alt="Screen Shot 2021-07-04 at 11 30 28 AM" src="https://user-images.githubusercontent.com/895923/124391273-cbbca680-dcbd-11eb-8cc6-030659b4bc9e.png">


Either way, for _greeting.ts_ the bare imports stay, well bare.  👀 

----

Let me know if you need more info or if this should be its own issue.  🙏

**Comment by LarsDenBakker** [comment]
@thescientist13 that's a different issues, you need to serve your files as JS. This might work:

```js
        return {
          body: result.outputText,
          type: 'js'
        };
```

but it might be that your plugin runs after node resolve. So then you might need to add a mimeTypes option to your config instead:

```js
{
  mimeTypes: {
    '**/*.ts': 'js',
},
```

**Comment by thescientist13** [comment]
Thanks @LarsDenBakker , that did it!  🙌

**Comment by abdonrd** [comment]
@tjenkinson is working on a fix for this (https://github.com/rollup/plugins/pull/921)! Thanks!

**Comment by zenwork** [comment]
> ```js
> return {
>           body: result.outputText,
>           type: 'js'
>         };
> ```

for anyone looking for a more complete quick workaround example:
```js
if (context.path.includes('node_modules/lit') && context.path.endsWith('.ts')) {
      const path = `${__dirname}/../../..${context.request.url}`.replace(/\.ts$/,'.js')
      console.log(path)
      const body = fs.readFileSync(path)
      return { body, type: 'js' };
    }
```

**Comment by abdonrd** [comment]
https://github.com/rollup/plugins/pull/921 has been solved and `@rollup/plugin-node-resolve@13.0.4` published!

**Comment by abdonrd** [comment]
@LarsDenBakker can you merge this https://github.com/modernweb-dev/web/pull/1433 to see if fix it? Thanks!

**Comment by abdonrd** [comment]
Fixed in `@web/dev-server@0.1.18`! 👏

**Comment by WickyNilliams** [comment]
Hey. What was the reason for rolling #1433 back in #1569?

I am experiencing this issue where an import of a `.js` file ends up being replaced with an import for a non-existent `.ts` file when served by the dev-server. 

Forcing my local install to use `@rollup/plugin-node-resolve@13.0.4` (by hand editing my package-lock.json) fixes the problem. So it seems this all is good in latest releases
